service/dap/server_test.go

```diff
@@ -1669,6 +1669,87 @@ func TestSetBreakpoint(t *testing.T) {
 	})
 }
 
+// TestLaunchSubstitutePath sets a breakpoint using a path
+// that does not exist and expects the substitutePath attribute
+// in the launch configuration to take care of the mapping.
+func TestLaunchSubstitutePath(t *testing.T) {
+	runTest(t, "loopprog", func(client *daptest.Client, fixture protest.Fixture) {
+		localDirPath := filepath.Join(string(filepath.Separator), "path", "that", "does", "not", "exist")
+		if runtime.GOOS == "windows" {
+			localDirPath = "C:" + localDirPath
+		}
+		localFilePath := filepath.Join(localDirPath, "loopprog.go")
+		debuggerDirPath := filepath.Dir(fixture.Source)
+
+		runDebugSessionWithBPs(t, client, "launch",
+			// Attach
+			func() {
+				client.LaunchRequestWithArgs(map[string]interface{}{"mode": "exec", "program": fixture.Path, "stopOnEntry": false, "substitutePath": []map[string]string{{"from": localDirPath, "to": debuggerDirPath}}})
+			},
+			// Set breakpoints
+			localFilePath, []int{8},
+			[]onBreakpoint{{
+				// Stop at line 8
+				execute: func() {
+					handleStop(t, client, 1, "main.loop", 8)
+				},
+				disconnect: true,
+			}})
+	})
+}
+
+// TestAttachSubstitutePath sets a breakpoint using a path
+// that does not exist and expects the substitutePath attribute
+// in the launch configuration to take care of the mapping.
+func TestAttachSubstitutePath(t *testing.T) {
+	if runtime.GOOS == "freebsd" {
+		t.SkipNow()
+	}
+	if runtime.GOOS == "windows" {
+		t.Skip("test skipped on windows, see https://delve.beta.teamcity.com/project/Delve_windows for details")
+	}
+	runTest(t, "loopprog", func(client *daptest.Client, fixture protest.Fixture) {
+		localDirPath := filepath.Join(string(filepath.Separator), "path", "that", "does", "not", "exist")
+		if runtime.GOOS == "windows" {
+			localDirPath = "C:" + localDirPath
+		}
+		localFilePath := filepath.Join(localDirPath, "loopprog.go")
+		debuggerDirPath := filepath.Dir(fixture.Source)
+
+		// Start the program to attach to
+		cmd := execFixture(t, fixture)
+
+		runDebugSessionWithBPs(t, client, "attach",
```

Comment:
How about putting shared code in a helper to avoid duplication and show how similar these tests are? Another option is to use the attach to test for bad substitute path where the mapping cannot be completed to see where that gets caughtt.

Reply:
Done. Will also add some tests for bad path mappings and ordering.